src/main/java/org/terasology/dynamicCities/buildings/BuildingManager.java

```diff
@@ -188,28 +182,47 @@ public void postBegin() {
 
     public Optional<GenericBuildingComponent> getRandomBuildingOfZoneForCulture(String zone, Rect2i shape, CultureComponent cultureComponent) {
         if (buildings.containsKey(zone)) {
-            int max = buildings.get(zone).size();
-            GenericBuildingComponent building;
+            // TODO: needs to be adapted if buildings have a specific spawn chance
+            List<GenericBuildingComponent> availableBuildings =
+                    buildings.get(zone).stream().filter(b -> cultureComponent.availableBuildings.contains(b.name)).collect(Collectors.toList());
+            int numberOfBuildingsForZone = availableBuildings.size();
+
+            GenericBuildingComponent selectedBuilding = null;
             int iter = 0;
-            do {
-                int index = rng.nextInt(max);
-                building = (GenericBuildingComponent) buildings.get(zone).toArray()[index];
+            int index;
+            do { // improve iteration by removing non-fitting buildings from the list
+                index = rng.nextInt(numberOfBuildingsForZone);
+                GenericBuildingComponent candidateBuilding = availableBuildings.get(index);
+                if (isFitting(shape, candidateBuilding)) {
+                    selectedBuilding = candidateBuilding;
+                } else {
+                    availableBuildings.remove(candidateBuilding);
+                    numberOfBuildingsForZone--;
+                }
                 iter++;
-            } while ((!isFitting(shape, building) || !cultureComponent.availableBuildings.contains(building.name)) && iter < 100);
-            if (iter >= 55) {
-                building = entityManager.getComponentLibrary().copy(findBiggestFittingBuilding(shape, zone));
-                building.isScaledDown = true;
+            } while (!availableBuildings.isEmpty() && selectedBuilding == null && iter < 100);
+
+            if (selectedBuilding == null) { // no building was found - try to find alternative
+                GenericBuildingComponent biggestFittingBuilding = findBiggestFittingBuilding(shape, zone);
+                if (biggestFittingBuilding != null) {
+                    selectedBuilding = entityManager.getComponentLibrary().copy(biggestFittingBuilding);
+                    selectedBuilding.isScaledDown = true;
+                }
+            }
+
+            if (selectedBuilding != null) {
+                logger.info("Found building \"{}\" for zone \"{}\" and size ({}, {})", selectedBuilding.name, zone, shape.width(), shape.height());
+                return Optional.of(entityManager.getComponentLibrary().copy(selectedBuilding));
```

Comment:
Currently the warnings produce more output than the info part :wink: But you are right, this is more of a debug information...